libraries/AP_NavEKF/AP_NavEKF_Source.cpp

```diff
@@ -0,0 +1,492 @@
+void AP_NavEKF_Source::setPosVelYawSource(uint8_t source_set_idx)
+{
+    // ensure init has been run
+    init();
+
+    // sanity check source idx
+    if (source_set_idx >= AP_NAKEKF_SOURCE_SET_MAX) {
+        return;
+    }
+
+    _active_source_set.posxy = (SourceXY)_source_set[source_set_idx].posxy.get();
+    _active_source_set.velxy = (SourceXY)_source_set[source_set_idx].velxy.get();
+    _active_source_set.posz = (SourceZ)_source_set[source_set_idx].posz.get();
+    _active_source_set.velz = (SourceZ)_source_set[source_set_idx].velz.get();
+    _active_source_set.yaw = (SourceYaw)_source_set[source_set_idx].yaw.get();
+}
+
+// true/false of whether velocity source should be used
+bool AP_NavEKF_Source::useVelXYSource(SourceXY velxy_source) const
+{
+    if (velxy_source == _active_source_set.velxy) {
+        return true;
+    }
+
+    // check for fuse all velocities
+    if (_options.get() & (uint16_t)(SourceOptions::FUSE_ALL_VELOCITIES)) {
+        for (uint8_t i=0; i<AP_NAKEKF_SOURCE_SET_MAX; i++) {
+            if (getVelXYSourceByIndex(i) == velxy_source) {
+                return true;
+            }
+        }
+    }
+
+    // if we got this far source should not be used
+    return false;
+}
+
+bool AP_NavEKF_Source::useVelZSource(SourceZ velz_source) const
+{
+    if (velz_source == _active_source_set.velz) {
+        return true;
+    }
+
+    // check for fuse all velocities
+    if (_options.get() & (uint16_t)(SourceOptions::FUSE_ALL_VELOCITIES)) {
+        for (uint8_t i=0; i<AP_NAKEKF_SOURCE_SET_MAX; i++) {
+            if (getVelZSourceByIndex(i) == velz_source) {
+                return true;
+            }
+        }
+    }
+
+    // if we got this far source should not be used
+    return false;
+}
+
+// true if a velocity source is configured
+bool AP_NavEKF_Source::haveVelZSource() const
+{
+    if (_active_source_set.velz != SourceZ::NONE) {
+        return true;
+    }
+
+    // check for fuse all velocities
+    if (_options.get() & (uint16_t)(SourceOptions::FUSE_ALL_VELOCITIES)) {
+        for (uint8_t i=0; i<AP_NAKEKF_SOURCE_SET_MAX; i++) {
+            if (getVelZSourceByIndex(i) != SourceZ::NONE) {
+                return true;
+            }
+        }
+    }
+
+    // if we got this far no velocity z source has been configured
+    return false;
+}
+
+// align position of inactive sources to ahrs
+void AP_NavEKF_Source::align_inactive_sources()
+{
+    // align visual odometry
+#if HAL_VISUALODOM_ENABLED
+    bool posxy_could_use_extnav = false;
+    bool posz_could_use_extnav = false;
+
+    for (uint8_t i=0; i<AP_NAKEKF_SOURCE_SET_MAX; i++) {
+        posxy_could_use_extnav |= (getPosXYSourceByIndex(i) == SourceXY::EXTNAV);
+        posz_could_use_extnav |= (getPosZSourceByIndex(i) == SourceZ::EXTNAV);
+    }
+
+    const bool align_posxy = posxy_could_use_extnav && ((getPosXYSource() == SourceXY::GPS) || (getPosXYSource() == SourceXY::BEACON));
```

Comment:
because the position source could also be set to NONE which is what we would expect if the user was using optical flow or wheel encoders.  We can either provide a list of sources that we shouldn't align our position to (EXTNAV, NONE) or a list that we should align to (GPS, BEACON).  I'm not sure either is better than the other when it comes to maintenance.

Reply:
Decided not to act on this for the time being.  If a new position source is added this code may need to be modified - we'll attack that at the time.